doc/nforenum.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" Please adjust this date whenever revising the manpage.
.Dd August 05, 2010
.Dt NFORENUM 1
.Sh NAME
.Nm NFORenum
.Nd A tool for checking NFO code for errors, and for beautifying that code
.Sh SYNOPSIS
.Nm nforenum
.Op options
.Op file [file2 [file3 [...]]]
.Sh OPTIONS
.Bl -tag
.It Fl -comments Ns = Ns Ar type | Fl c Ar type
<type> is one character, either /, ;, or #, and specifies the comment
style that NFORenum will use. This will not change the header, because
grfcodec requires that the header be commented in C++-style.
.It Fl -data Ns [= Ns Ar dir ] | Fl D Ns [ Ar dir ]
If <dir> is specified, look for the .nforenum directory in <dir>, and
create it if not found.
.Pp
If <dir> is not specified, report the location of the .nforenum directory.
In either case, eliminate the 5-second wait used to ensure that those
not running NFORenum from a command line can see the directory-created
message.
.Pp
Default: Look for  the .nforenum directory in the current directory, and
then in the environment variable
.Ev HOME Ns
, if defined. If not found attempt to create in
.Ev HOME Ns
, then in .
.It Fl -force | Fl -f
Forces processing of files that do not look like NFO files.
The default is to treat such files as if they specified a too-high info version.
.It Fl -help | Fl -h | Fl -?
Display the help message.
.It Fl -lock
Locks the current comment command state. Commands will continue to be
parsed as normal (so NOPRESERVE will be honored, @@DIFF will be
removed, &c.) but their changes (such as turning on the diff-assister)
will not be honored.
.It Fl -no-replace | Fl -keep-old | Fl k
Do not replace the old NFO file; write new file to file.new.nfo.
Default: Use file[.nfo].new as temporary, rename it to file[.nfo]
when done.
.It Fl -auto-correct | Fl a
Perform some rudimentary automatic correction of incorrect pseudo
sprites. This may be specified twice to enable the corrections that
are more likely to be incorrect.
See the README for detailed information on the auto-correcter.
.El

The following options cause NFORenum to behave as if all files started with
the associated command. The readme has full details on the comment commands.
Options associated with comment commands that require additional information
take that information as an argument. With the exception of -L/--let, the
options to the command line versions are case insensitive.
"ON" and "OFF" may be specified with "+" and "-", respectively.
.Bl -tag -width 24n
.It Fl -beautify | Fl b
@@BEAUTIFY
.It Fl -diff | Fl d
@@DIFF
.It Fl -let | Fl L
@@LET
.It Fl -lint | Fl l
@@LINT
.It Fl -preserve-messages | Fl p
@@PRESERVEMESSAGES
.It Fl -real-sprites | Fl r
@@REALSPRITES
.It Fl -use-old-nums | Fl o
@@USEOLDSPRITENUMS
.It Fl -warning-disable | Fl w
@@WARNING DISABLE
.It Fl -warning-enable | Fl W
@@WARNING ENABLE
-w and -W (and their long counterparts) also accept a comma-separated
list of messages, all of which will be ENABLEd or DISABLEd.

.Sh SEE ALSO
The documentation in
.Pa /usr/share/doc/nforenum/
.Sh AUTHOR
nforenum was written by Dale McCoy, with portions by Dan Masek, Thijs
Marinussen and Remko Bijker.
.Pp
This manual page was written by Remko Bijker.


Changelog
Changes in
- (devel) Detect UPX, fix some race issues in the Makefile, disable
  optimization if GCC 4.5.0 is detected (to work around an optimization bug),
  fix version.h generation if there is a version.h in /usr/include.
- (devel) Fix a number of compiler warnings.
- Clamp the arguments for -w and -h to 65535 as that is the maximum size of a
  PCX.
- Silence progress-meter output when stdout is not a console.
- Unify warnings and errors so they go to stderr and are prefixed by "Warning:"
  or "Error:".
- Allow hex escapes that have bit 31 set.
- After cropping a sprite, compress and write the correct number of bytes.
- Add -X to print PCX sprite numbers in hex
- (devel) Many compilation issues with different compilers and architectures.
- Fix bug in UTF-8 generation for escapes \U800..\U7FFF.
- Fix U+E0xx characters being output quoted.
- (devel) By Tron: code cleanup, make C functions and variables static where
  possible, remove dead and debugging code.
- Require two slashes before dismissing anything as a C++ style comment.
- Fix reading of real sprites with Unixy absolute paths.
- Check for pixels with palette index of 255 when encoding.
- Add -q to silence warnings.
- Add -c to automatically crop extraenous transparent blue pixels.
- Add escapes for the new action 2 operations 0E, 0F, 11, 12, 13, 14, 15, 16
- Prevent quotes and escapes in sprite 0.
- Add escapes for long-format intro dates.

Changes in 0.9.10 (r819)
- (bugfix) Correct reading of \7c, \7C, \D| and \D&. (DaleStan)

Changes in 0.9.9 (r653)
- By Patchman:
  - Check for sprites that extend beyond the end of the PCX file.
  - Update versioning system to show last-comitted revision for non-release
    builds.
- By DaleStan:
  - (devel) Add release targets to the Makefiles
  - Check the palette of the PCX file before encoding from it.
    (Override a failed check with -f)
  - Introduce several backslash escape sequences; see the wiki for further
    information.
    (disable with -x or -xx)
  - Do not apply colormap to sprites that like character glyphs.
    (Apply colormap to all sprites with -M instead of -m)
  - Enable reading of GIMP palette files with -p.
  - (devel) Terse build mode; get the verbose mode with make V=1.
  - (devel) Touch version.h if and only if it changes.

Changes in 0.9.8 (r197)
- By DaleStan:
  - (devel) grfcodec (but not grfdiff or grfmerge) can now be built in
    MS Visual Studio 2003.
  - Slightly change the plaintext quoting algorithm.
  - (devel) Change several memcpy()s to pointer assignments.
  - (devel) Plug several memory leaks
  - (devel) Source code and Makefile changes to grfdiff and grfmerge build
    with both Makefiles, on both Cygwin and Linux.
  - Make code gcc 4 -Wall clean.
  - Use NFORenum's NFO reading engine:
    - Sprite numbers are now optional, and need not be correct if present.
    - Sprite lengths need not be correct. Set to 0 to supress the incorrect
      length warnings.
    - Real sprites are checked for format compliance.
  - Make 0-byte pseudo sprites an error.
  - (bugfix) Prevent binary includes from including directories on OSes that
    would othewise allow it.
  - (bugfix) It is now possible for a real sprite to start above the
    preceeding sprite (the infamous decreasing ypos bug)
  - grfcodec now exits with 0 on success.

Changes in 0.9.7
- support including binary files using "  4 ** data/dingdong.wav" syntax
- by Oskar: support // and # comments within pseudo-sprites
- fixed incorrect error message while encoding if .nfo file could not be read
  (now it reports "Cannot read %s" instead of Creating sprites/: File exists")
- made code -Wall clean
- remove myalloc stuff

Changes in 0.9.6
- by pasky:
  * made it possible to compile grfcodec on Linux
  * made compatible with gcc 3.x
  * when decoding, text strings are written in plain text (disable with -t)
  * supports plain text in .nfo files
  * fixed bug when first encoding a grf file wouldn't rename .new to .grf

Changes in 0.9.5
- self-extracting executables now work on Win2K etc., which doesn't pass
  the full path to the C library
- GRFMerge will just skip .grf files it can't find
- making sprites larger doesn't work if they're at the right edge; complains
  that ofsx too large; instead just use background colour
- GRFDiff doesn't like grfdiff -l 1254 ... (single value in -l), complains
  that there were no differences and makes no output
- allow GRFMerge to read GRFDiff's .exe files, both for -l and other options

Changes in 0.9.4
- fixed yet another crash while compressing
- sprites no longer need correct numbers in the .NFO file; just use "-1" for
  each number and it will be assumed to be correct

Changes in 0.9.3
- fixed crash due to buffer overrun while compressing
- slightly improved compression efficiency and speed
- GRFCodec now handles spaces in file/directory names properly
- GRFCodec now supports comments in .NFO files (from //, ; or # to end of line)
- speed up PCX processing

Changes in 0.9.2
- fixed crash if pcx file not found
- fixed crash if nfo file has no version info header

Changes in 0.9.1
- bug fix by Marcin: wrong order of test in encode function

Changes in release April 20 2002
- -p now actually uses the parameter you give it

Changes in release Dec 03 2001
- made -p option a valid option
- improve compression speed even more using assembly optimizations
- compression is now on by default, disable with -c-
- fixed bug where encoding would fail if the grf file didn't exist already
- fixed bug where decoding would fail creating the sprites directory
- corrected some mistakes in grf.txt (checksum, and initial offsets for tiles)

Changes in release March 26 2001
- fix bug that didn't flush last row if it only contained a single sprite
- fix bug with no height given (would make PCX have height -1)
- show percentage done while decoding
- fix compression algorithm
- improve compression performance
- try without compression if data is inconsistent after decompressing our

src/nforenum.cpp
/*
 * renum.cpp
 * Renumbers and lints an NFO file specified on the commandline or piped
 * in via standard input.
 *
 * Return values:
 * See README[.txt]
 *
 * Copyright 2004-2007,2009 by Dale McCoy.
 * dalestan@gmail.com
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include<iostream>
#include<string>
#include<sstream>
#include<iomanip>
#include<fstream>
#include<cerrno>
#include<cassert>
#include<cstdlib>

using namespace std;

#include"nforenum.h"
#include"getopt.h"
#include"globals.h"
#include"inlines.h"
#include"sanity.h"
#include"command.h"
#include"messages.h"
#include"inject.h"
#include"pseudo.h"
#include"help.h"
#ifdef _WIN32
#   include "win32.h"
#endif

#include "mapescapes.h"

nfe_map nfo_escapes;

#ifndef _MSC_VER
//Cygwin's GCC #defines __cdecl, but other GCCs do not
//#undef it to prevent GCC from warning on the #define,
//then #define it to prevent errors on main's definition.
#undef __cdecl
#define __cdecl
#endif

int process_file(istream&);
void output_buffer(const string&,bool,int);
bool verify_real(string&);

static int _retval=EOK;
static int _force=0;

void SetCode(int x){_retval=max(_retval,x);}
void doexit(){exit(_retval);}

//The VS project file specifies __stdcall as the default convention,
//so main must be explicitly __cdecl'd
int __cdecl main(const int argc,char**argv){
	string infilename,outfilename,bakfilename,basename;
	int result,longind,opt=0,replace=1;
	static const option optlist[]={
		{"data",optional_argument,NULL,'D'},
		{"comments",required_argument,NULL,'c'},
		{"force",no_argument,&_force,1},
		{"lock",no_argument,NULL,256},
		{"no-replace",no_argument,&replace,0},
		{"keep-old",no_argument,&replace,0},
		{"help",no_argument,NULL,'?'},
		{"help",no_argument,NULL,'h'},
		{"auto-correct",no_argument,NULL,'a'},
		{"beautify",required_argument,NULL,'b'},
		{"diff",no_argument,NULL,'d'},
		{"extentions",required_argument,NULL,'e'},
		{"let",required_argument,NULL,'L'},
		{"lint",optional_argument,NULL,'l'},
		{"preserve-messages",no_argument,NULL,'P'},
		{"real-sprites",required_argument,NULL,'r'},
		{"use-old-nums",required_argument,NULL,'o'},
		{"warning-disable",required_argument,NULL,'w'},
		{"warning-enable",required_argument,NULL,'W'},
		{NULL,0,0,0}
	};
	pOut=argc==1?&cerr:&cout;
	IssueMessage(0,STARTUP);
	ifstream fin;
	ofstream fout;
	while(argc>1){
		if(opt!=EOF)opt=getopt_long(argc,argv,"D::kv?hc:fa" "dL:l:pw:W:r:b:e:",optlist,&longind);
		switch(opt){
		case 0:continue;
		case 'v':
			/* The version is already printed. */
			return 0;
		case'D':
			if(optarg)datadir=optarg;
			dosleep=false;
			continue;
		case'k':replace=0;continue;
		case'c':
			switch(*optarg){
			case'#':COMMENT_PREFIX="#";break;
			case';':COMMENT_PREFIX=";";break;
			case'/':COMMENT_PREFIX="//";
			}
			continue;
		case'?':
		case'h':
			ShowHelp();
			return 0;
		case'f':_force=1;continue;
		case'a':
			_autocorrect++;
			if(!GetState(BEAUTIFY)){
				optarg=(char*)"convertonly+";
				CLCommand('b');
			}
			continue;
		case EOF:
			if(optind==argc)doexit();
			basename=argv[optind++];break;
		default:
			if(!CLCommand(opt))
				IssueMessage(0,BAD_CL_ARG,opt,optarg);
			continue;
		}
		pNfo=&fout;
		bakfilename=basename+bak_ext;
		fin.open((infilename=basename).c_str());
		if(!fin.is_open()){
			fin.open((infilename=basename+nfo_ext).c_str());
			bakfilename=basename+nfo_ext+bak_ext;
		}
		if(fin.is_open()){
			fout.open((outfilename=(replace?basename+new_ext:basename+new_ext+nfo_ext)).c_str());
		}else{
			bakfilename=dirname+(basename+bak_ext);
			fin.open((infilename=dirname+basename).c_str());
			if(!fin.is_open()){
				bakfilename=dirname+(basename+nfo_ext+bak_ext);
				fin.open((infilename=dirname+basename+nfo_ext).c_str());
			}
			if(fin.is_open()){
				fout.open((outfilename=dirname+(replace?basename+new_ext:basename+new_ext+nfo_ext)).c_str());
			}else{
				IssueMessage(0,NO_INPUT_FILE,basename.c_str());
				SetCode(EFILE);
				continue;
			}
		}
		if(!fout.is_open()){
			IssueMessage(0,NO_OUTPUT_FILE,outfilename.c_str(),basename.c_str());
			SetCode(EFILE);
			fin.close();
			fout.clear();
			continue;
		}
		fin.clear();
		reset_sanity();
		reset_commands();
		_grfver=0;
		IssueMessage(0,PROCESSING_FILE,basename.c_str());
		result=process_file(fin);
		fin.close();
		fout.close();
		if(result){
#if defined(_MSC_VER) && (_MSC_VER >= 1400)
// unlink is deprecated in MSVS 8.0+
			_unlink(outfilename.c_str());
#else
			unlink(outfilename.c_str());
#endif
		}else if(replace){
			if(rename(infilename.c_str(),bakfilename.c_str())&&errno==EEXIST){
#if defined(_MSC_VER) && (_MSC_VER >= 1400)
// unlink is deprecated in MSVS 8.0+
				if(_unlink(infilename.c_str()))
#else
				if(unlink(infilename.c_str()))
#endif
				{
					IssueMessage(0,DELETE_FAILED,infilename.c_str(),errno);
					perror(NULL);
					SetCode(EFILE);
				}
			}
			if(rename(outfilename.c_str(),infilename.c_str())){
				IssueMessage(0,REPLACE_FAILED,infilename.c_str(),outfilename.c_str(),errno);
				perror(NULL);
				SetCode(EFILE);
			}
		}
		IssueMessage(0,PROCESSING_COMPLETE);
	}
	pNfo=&cout;
	IssueMessage(0,PROCESSING);
	process_file(cin);
	IssueMessage(0,PROCESSING_COMPLETE);
	return _retval;
}

#define flush_buffer()\
	if(true) {\
		if(buffer!="") {\
			output_buffer(buffer,isPatch,oldspritenum);\
			buffer="";\
		}\
		oldspritenum=temp;\
	} else\
		(void(0))

#define SetVersion(x)\
	(NFOversion=max((x),NFOversion))

static bool hasHeader;
int NFOversion;

bool TrySetVersion(int x){
	if(hasHeader&&NFOversion>=x)return true;
	if(hasHeader&&NFOversion<=6&&x>=7)return false;
	SetVersion(x);
	return true;
}

string smash(const string&,int&);

int process_file(istream&in){
	NFOversion=4;
	string sprite,datapart,buffer;
	inject_into(in);
	nfo_escapes.clear();
	vector<string> extra_lines;

	if(string(" \t\n*0123456789/;#*").find((char)in.peek())==NPOS){
		IssueMessage(0,APPARENTLY_NOT_NFO);
		if(!_force){
			IssueMessage(0,SKIPPING_FILE);
			SetCode(EPARSE);
			return-1;
		}
	}

//This section is a rewrite of lines 35-50 or thereabouts of info.cc
//from grfcodec: http://www.ttdpatch.net/grfcodec
//grfcodec is Copyright 2000-2003 Josef Drexler
	getline(in,sprite);//Info header or first sprite
	hasHeader=false;
	if (sprite.substr(0,3)=="// "){
		hasHeader=true;
		getline(in,sprite);//Info version (first sprite for ancient NFO versions)
		if(sscanf(sprite.c_str(),"// (Info version %d)",&NFOversion)){
			if(NFOversion<4||NFOversion>7){
				IssueMessage(0,UNKNOWN_VERSION,NFOversion);
				if(NFOversion>7){
					IssueMessage(0,SKIPPING_FILE);
					SetCode(EPARSE);
					return-1;
				}
				IssueMessage(0,PARSING_FILE);
			}
			if(NFOversion>2)
				getline(in,sprite);//Format line
			if (NFOversion>6) {
				while (strncmp(sprite.c_str(), "// Format: ", 11)) {
					if (!strncmp(sprite.c_str(), "// Escapes: ", 12)) {	// Actually, that was an escapes line. Read it.
						istringstream esc(sprite);
						string str;
						int byte = 0;
						esc.ignore(12);
						while (esc>>str)
							if (str == "=") byte--;
							else if (str[0] == '@')
								byte = strtol(str.c_str()+1,NULL,16);
							else nfo_escapes.insert(nfe_pair(str, byte++));
					} else if (strncmp(sprite.c_str(), "// ",3)) { // EEEP! No "Format:" line in this file!
						IssueMessage(0, APPARENTLY_NOT_NFO);
						SetCode(EPARSE);
						return -1;
					} else extra_lines.push_back(sprite); // store unknown lines

					getline(in,sprite);	// Try again to skip "Format: " line
				}
				// Now remove all defaults. This serves two purposes:
				// 1) Prevent incorrectly specified defaults from causing problems later.
				// 2) Allow the beautifier to select custom escapes over built-ins.
				foreach(const esc& e, escapes)
					nfo_escapes.left.erase(e.str+1);
			}
		}else{
			IssueMessage(0,UNKNOWN_VERSION,1);
			IssueMessage(0,PARSING_FILE);
		}
		getline(in,sprite);//first sprite
	}


	int temp=-1,size,oldspritenum=-1;
	_spritenum=(unsigned)-1;
	string::size_type firstnotpseudo;
	bool isPatch=false;
	ostringstream outbuffer;
	ostream*real_out=pNfo;
	pNfo=&outbuffer;
	SetVersion(4);
	AutoConsoleMessages();
	while(true){
		//IssueMessage(0,SPRITE,spritenum+1,sprite.c_str());
		istringstream spritestream(sprite);
		eat_white(spritestream);
		if(spritestream.peek()==EOF) {
            buffer+='\n';
		} else if(is_comment(spritestream)){
			if(is_command(sprite)) {
				flush_buffer();
            }
			if(parse_comment(sprite)) {
				buffer+=sprite+'\n';
            }
		} else {//sprite
			if(!eat_white(spritestream>>temp)){
				spritestream.clear();
				temp=-1;
			}
			if(spritestream.peek()=='*'){
				if(spritestream.ignore().peek()=='*'){
					SetVersion(6);
					getline(eat_white(spritestream.ignore()),datapart);
					flush_buffer();
					if(isPatch) {
						check_sprite(INCLUDE);
					} else {
						IssueMessage(ERROR,UNEXPECTED,BIN_INCLUDE);
					}
					_spritenum++;
					(*pNfo)<<setw(5)<<spritenum()<<" **\t "<<datapart<<'\n';
				}else{
					(spritestream>>size).clear();
					eat_white(spritestream);
					flush_buffer();
					if(_spritenum==(uint)-1){
						isPatch=true;
						if(size!=4 && size!=0)
							outbuffer<<COMMENT_PREFIX<<sprite<<endl;
						_spritenum=0;
					}else{
						getline(spritestream,buffer);
						buffer+='\n';
					}
				}
			}else{
				getline(spritestream,datapart);
				firstnotpseudo=datapart.find_first_not_of(VALID_PSEUDO);
				if(!spritestream || firstnotpseudo==NPOS
							|| datapart[firstnotpseudo]=='"'
							|| (datapart[firstnotpseudo]=='\\'?TrySetVersion(7):false)
							|| is_comment(datapart,firstnotpseudo)) {
					if(PseudoSprite::MayBeSprite(buffer)) {
						buffer+=sprite+'\n';
					} else {
						IssueMessage(0,NOT_IN_SPRITE,_spritenum+1);
						buffer+="//"+sprite+'\n';
					}
				}else{
					_spritenum++;
					if(verify_real(datapart)){
						_spritenum--;
						flush_buffer();
						_spritenum++;
						if(isPatch)check_sprite(REAL);
						(*pNfo)<<setw(5)<<spritenum()<<' '<<datapart<<endl;
					}else{
						buffer+="//"+sprite+'\n';
						SetCode(EPARSE);
						IssueMessage(0,PARTIAL_PARSE_FAILURE,_spritenum--);
					}
				}
			}
		}
		if(peek(in)==EOF){
			flush_buffer();
			(*real_out)<<NFO_HEADER(NFOversion);
			if (NFOversion > 6) {
					// (re)insert default escapes
					foreach(const esc& e, escapes)
						nfo_escapes.insert(nfe_pair(e.str+1, e.byte));
					(*real_out)<<"// Escapes:";
					int oldbyte = -1;

					for (int act = 0; act < 255; act++) {
						foreach (const nfe_rpair& p, nfo_escapes.right) {
							if (p.second[0] != act) continue;

							if (p.first == oldbyte) {
								(*real_out)<<" =";
								--oldbyte;
							} else if (p.first < oldbyte) {
								(*real_out)<<"\n// Escapes:";
								oldbyte = -1;
							}
							while (++oldbyte != p.first)
								(*real_out)<<" "<<nfo_escapes.right.begin()->second;
							(*real_out)<<" "<<p.second;
						}
					}
					(*real_out)<<"\n";
				for (uint i=0; i<extra_lines.size(); i++)
					(*real_out)<<extra_lines[i]<<"\n";
			}
			(*real_out)<<NFO_FORMAT;
			if(isPatch)(*real_out)<<"    0 * 4\t "<<mysprintf("%8x\n",GetState(DIFF)?0:_spritenum);
			(*real_out)<<outbuffer.str();
			pNfo=real_out;
			final_sanity();
			return 0;
		}
		inj_getline(in,sprite);
	}
}

bool verify_real(string&data){
	string::size_type loc=NPOS;
	string udata=UCase(data);
	while(true){
		loc=udata.find(".PCX",loc+1);
		if(loc==NPOS){
			IssueMessage(0,REAL_NO_FILENAME);
			return false;
		}
		if(isspace(data[loc+4]))break;
	}
	string name=data.substr(0,loc+4);
	int var_list[7]={0,0,1,1,1,0,0},
		&xpos=var_list[0],
		&ypos=var_list[1],
		&comp=var_list[2],
		&ysize=var_list[3],
		&xsize=var_list[4],
		&xrel=var_list[5],
		&yrel=var_list[6];
	const char*const var_names[7]={"xpos","ypos","comp","ysize","xsize","xrel","yrel"};
	string meta=data.substr(loc+5);
	string::size_type offs=NPOS;
	uchar state=0; // bitmask of calculated variables
	int var;
	while((var=sscanf(meta.c_str(),"%d %d %2x %d %d %d %d",&xpos,&ypos,&comp,&ysize,&xsize,&xrel,&yrel))!=7){
		if(RPNOFF()||
			//Calculation on comp, already tried to calculate this datum, or no more parens->broken sprite
			var==2||state&(1<<var)||(offs=meta.find('('))==NPOS||
			//If OK so far, calculate and store datum. If calc fails, offs == NPOS; sprite is broken
			(var_list[var]=DoCalc(meta,offs),offs==NPOS)){
			if(COMMENTON()||RPNON()||(RPNOFF()&&meta.find('(')==NPOS))IssueMessage(0,REAL_MISSING_DATA,var_names[var]);
			return COMMENTOFF();
		}
		state|=1<<var;//Mark calculation
		const char* rest(meta.c_str()+offs);
		switch(var) {
		case 0: meta=mysprintf("%d%t",xpos,rest); break;
		case 1: meta=mysprintf("%d %d%t",xpos,ypos,rest); break;
		case 3: meta=mysprintf("%d %d %2x %d%t",xpos,ypos,comp,ysize,rest); break;
		case 4: meta=mysprintf("%d %d %2x %d %d%t",xpos,ypos,comp,ysize,xsize,rest); break;
		case 5: meta=mysprintf("%d %d %2x %d %d %d%t",xpos,ypos,comp,ysize,xsize,xrel,rest); break;
		default: meta=mysprintf("%d %d %2x %d %d %d %d%t",xpos,ypos,comp,ysize,xsize,xrel,yrel,rest); break;
		}
	}
	if(state)data=data.substr(0,loc+5)+meta;
	if(xpos<0)IssueMessage(ERROR,REAL_VAL_TOO_SMALL,XPOS,0);
	if(ypos<0)IssueMessage(ERROR,REAL_VAL_TOO_SMALL,YPOS,0);
	if(!(comp&1)||(comp&0x4B)!=comp)IssueMessage(comp==0xFF?ERROR:WARNING1,REAL_BAD_COMP,comp);
	if(xsize<1)IssueMessage(ERROR,REAL_VAL_TOO_SMALL,XSIZE,1);
	else if(xsize>0xFFFF)IssueMessage(ERROR,REAL_VAL_TOO_LARGE,XSIZE,0xFFFF);
	if(ysize<1)IssueMessage(ERROR,REAL_VAL_TOO_SMALL,YSIZE,1);
	else if(ysize>0xFF)IssueMessage(ERROR,REAL_VAL_TOO_LARGE,YSIZE,0xFF);
	if(xsize*ysize>0xFFFF)IssueMessage(ERROR,REAL_SPRITE_TOO_LARGE);
	if(xrel<-32768)IssueMessage(ERROR,REAL_VAL_TOO_SMALL,XREL,-32768);
	else if(xrel>32767)IssueMessage(ERROR,REAL_VAL_TOO_LARGE,XREL,32767);
	if(yrel<-32768)IssueMessage(ERROR,REAL_VAL_TOO_SMALL,YREL,-32768);
	else if(yrel>32767)IssueMessage(ERROR,REAL_VAL_TOO_LARGE,YREL,32767);
	return true;
}

void output_buffer(const string&sprite,bool isPatch,int spriteno){
	PseudoSprite data(sprite,spriteno);
	if(data.IsValid()){
		_spritenum++;
		if(isPatch)check_sprite(data);
		else{
			data.SetAllHex();
			data.SetEol(32,1);
			data.SetEol(64,1);
			data.SetEol(96,1);
			data.SetEol(128,1);
			data.SetEol(160,1);
			data.SetEol(192,1);
			data.SetEol(224,1);
		}
	}
	(*pNfo)<<data;
}
